File: agent-ops/src/services/githubMcpService.ts
/**
 * GitHub MCP Service - Frontend service for GitHub MCP integration
 * Connects to backend GitHub MCP APIs and provides workflow management
 */

export interface GitHubMCPTool {
  name: string;
  description: string;
  parameters: Record<string, any>;
  required_parameters: string[];
}

export interface GitHubRepository {
  id: number;
  name: string;
  full_name: string;
  private: boolean;
  html_url: string;
  description?: string;
  language?: string;
  default_branch: string;
  updated_at: string;
  open_issues_count: number;
}

export interface GitHubWorkflow {
  workflow_id: string;
  workflow_type: string;
  status: 'pending' | 'analyzing' | 'implementing' | 'reviewing' | 'awaiting_approval' | 'approved' | 'rejected' | 'completed' | 'failed';
  context: Record<string, any>;
  created_at: string;
  updated_at: string;
  steps: WorkflowStep[];
  approvals: Record<string, boolean>;
  ai_analysis?: Record<string, any>;
}

export interface WorkflowStep {
  step_id: string;
  type: string;
  description: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  timestamp: string;
  details?: Record<string, any>;
}

export interface GitHubMCPExecuteRequest {
  operation: string;
  params: Record<string, any>;
}

export interface GitHubMCPExecuteResponse {
  success: boolean;
  message: string;
  result: Record<string, any>;
  operation: string;
}

export interface WorkflowApprovalRequest {
  approved: boolean;
  reviewer: string;
  notes?: string;
}

class GitHubMCPService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = process.env.NODE_ENV === 'production' 
      ? '/api/github-mcp' 
      : 'http://localhost:8000/api/github-mcp';
  }

  /**
   * Get available GitHub MCP tools for Claude integration
   */
  async getAvailableTools(): Promise<GitHubMCPTool[]> {
    try {
      const response = await fetch(`${this.baseUrl}/tools`);
      if (!response.ok) {
        throw new Error(`Failed to fetch tools: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      console.error('Error fetching GitHub MCP tools:', error);
      throw error;
    }
  }

  /**
   * Execute a GitHub MCP operation
   */
  async executeOperation(request: GitHubMCPExecuteRequest): Promise<GitHubMCPExecuteResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/execute`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        throw new Error(`Failed to execute operation: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error executing GitHub MCP operation:', error);
      throw error;
    }
  }

  /**
   * List user repositories
   */
  async listRepositories(params: {
    visibility?: 'all' | 'public' | 'private';
    sort?: 'created' | 'updated' | 'pushed' | 'full_name';
    direction?: 'asc' | 'desc';
    per_page?: number;
  } = {}): Promise<GitHubRepository[]> {
    try {
      const response = await this.executeOperation({
        operation: 'list_repositories',
        params,
      });

      if (response.success) {
        return response.result.repositories || [];
      } else {
        throw new Error(response.message);
      }
    } catch (error) {
      console.error('Error listing repositories:', error);
      throw error;
    }
  }

  /**
   * Get repository content
   */
  async getRepositoryContent(params: {
    owner: string;
    repo: string;
    path?: string;
    ref?: string;
  }): Promise<any> {
    try {
      const response = await this.executeOperation({
        operation: 'get_repository_content',
        params,
      });

      if (response.success) {
        return response.result.content;
      } else {
        throw new Error(response.message);
      }
    } catch (error) {
      console.error('Error getting repository content:', error);
      throw error;
    }
  }

  /**
   * Create a new branch
   */
  async createBranch(params: {
    owner: string;
    repo: string;
    branch_name: string;
    base_branch?: string;
  }): Promise<any> {
    try {
      const response = await this.executeOperation({
        operation: 'create_branch',
        params,
      });

      if (response.success) {
        return response.result;
      } else {
        throw new Error(response.message);
      }
    } catch (error) {
      console.error('Error creating branch:', error);
      throw error;
    }
  }

  /**
   * Create a pull request
   */
  async createPullRequest(params: {
    owner: string;
    repo: string;
    title: string;
    body: string;
    head: string;
    base: string;
    draft?: boolean;
  }): Promise<any> {
    try {
      const response = await this.executeOperation({
        operation: 'create_pull_request',
        params,
      });

      if (response.success) {
        return response.result;
      } else {
        throw new Error(response.message);
      }
    } catch (error) {
      console.error('Error creating pull request:', error);
      throw error;
    }
  }

  /**
   * Start Issue-to-PR workflow
   */
  async startIssueToprWorkflow(params: {
    owner: string;
    repo: string;
    issue_number: number;
    claude_api_key_name?: string;
    api_key_name?: string;
  }): Promise<{ workflow_id: string; status: string }> {
    try {
      const response = await fetch(`${this.baseUrl}/workflows/issue-to-pr?${new URLSearchParams({
        owner: params.owner,
        repo: params.repo,
        issue_number: params.issue_number.toString(),
        claude_api_key_name: params.claude_api_key_name || 'claude',
        api_key_name: params.api_key_name || 'github',
      })}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        throw new Error(`Failed to start workflow: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error starting issue-to-PR workflow:', error);
      throw error;
    }
  }

  /**
   * Start code review workflow
   */
  async startCodeReviewWorkflow(params: {
    owner: string;
    repo: string;
    pull_number: number;
    claude_api_key_name?: string;
    api_key_name?: string;
  }): Promise<{ workflow_id: string; status: string }> {
    try {
      const response = await fetch(`${this.baseUrl}/workflows/code-review?${new URLSearchParams({
        owner: params.owner,
        repo: params.repo,
        pull_number: params.pull_number.toString(),
        claude_api_key_name: params.claude_api_key_name || 'claude',
        api_key_name: params.api_key_name || 'github',
      })}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        throw new Error(`Failed to start workflow: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error starting code review workflow:', error);
      throw error;
    }
  }

  /**
   * Get all active workflows
   */
  async getActiveWorkflows(): Promise<GitHubWorkflow[]> {
    try {
      const response = await fetch(`${this.baseUrl}/workflows`);
      if (!response.ok) {
        throw new Error(`Failed to fetch workflows: ${response.statusText}`);
      }

      const data = await response.json();
      return data.workflows || [];
    } catch (error) {
      console.error('Error fetching active workflows:', error);
      throw error;
    }
  }

  /**
   * Get workflow status by ID
   */
  async getWorkflowStatus(workflowId: string): Promise<GitHubWorkflow | null> {
    try {
      const response = await fetch(`${this.baseUrl}/workflows/${workflowId}`);
      if (!response.ok) {
        if (response.status === 404) {
          return null;
        }
        throw new Error(`Failed to fetch workflow: ${response.statusText}`);
      }

      const data = await response.json();
      return data.workflow;
    } catch (error) {
      console.error('Error fetching workflow status:', error);
      throw error;
    }
  }

  /**
   * Approve or reject a workflow step
   */
  async approveWorkflowStep(workflowId: string, approval: WorkflowApprovalRequest): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/workflows/${workflowId}/approve`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(approval),
      });

      if (!response.ok) {
        throw new Error(`Failed to approve workflow: ${response.statusText}`);
      }

      const data = await response.json();
      return data.success;
    } catch (error) {
      console.error('Error approving workflow step:', error);
      throw error;
    }
  }

  /**
   * Cancel a workflow
   */
  async cancelWorkflow(workflowId: string): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/workflows/${workflowId}`, {
        method: 'DELETE',
      });

      if (!response.ok) {
        throw new Error(`Failed to cancel workflow: ${response.statusText}`);
      }

      const data = await response.json();
      return data.success;
    } catch (error) {
      console.error('Error cancelling workflow:', error);
      throw error;
    }
  }

  /**
   * Clean up completed workflows
   */
  async cleanupCompletedWorkflows(maxAgeHours: number = 24): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/workflows/cleanup`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ max_age_hours: maxAgeHours }),
      });

      if (!response.ok) {
        throw new Error(`Failed to cleanup workflows: ${response.statusText}`);
      }

      const data = await response.json();
      return data.success;
    } catch (error) {
      console.error('Error cleaning up workflows:', error);
      throw error;
    }
  }

  /**
   * Get service health status
   */
  async getHealthStatus(): Promise<{
    status: string;
    tools_available: number;
    active_workflows: number;
    features: string[];
  }> {
    try {
      const response = await fetch(`${this.baseUrl}/health`);
      if (!response.ok) {
        throw new Error(`Health check failed: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error checking GitHub MCP health:', error);
      throw error;
    }
  }

  /**
   * Subscribe to workflow updates (using Server-Sent Events or WebSocket)
   * For now, we'll implement polling, but this could be enhanced with real-time updates
   */
  async subscribeToWorkflowUpdates(
    workflowId: string,
    callback: (workflow: GitHubWorkflow) => void,
    pollInterval: number = 5000
  ): Promise<() => void> {
    const poll = async () => {
      try {
        const workflow = await this.getWorkflowStatus(workflowId);
        if (workflow) {
          callback(workflow);
        }
      } catch (error) {
        console.error('Error polling workflow status:', error);
      }
    };

    // Initial poll
    await poll();

    // Set up polling interval
    const intervalId = setInterval(poll, pollInterval);

    // Return cleanup function
    return () => {
      clearInterval(intervalId);
    };
  }
}

// Export singleton instance
export const githubMcpService = new GitHubMCPService();
export default githubMcpService; 

File: agent-ops/backend/app/services/github_mcp_service.py
"""
GitHub MCP Service - Core GitHub API integration with MCP tool format
"""
import asyncio
import base64
import json
import logging
from datetime import datetime
from typing import Dict, List, Any, Optional, Union
import aiohttp
from pydantic import BaseModel

from ..models.github_mcp_models import *
from ..config import settings

logger = logging.getLogger(__name__)


class GitHubMCPService:
    """Service for GitHub MCP operations with Claude integration"""
    
    def __init__(self):
        self.base_url = "https://api.github.com"
        self.sessions: Dict[str, aiohttp.ClientSession] = {}
        self.mcp_tools: List[MCPToolDefinition] = []
        self._initialize_mcp_tools()
    
    def _initialize_mcp_tools(self):
        """Initialize MCP tool definitions for Claude integration"""
        self.mcp_tools = [
            MCPToolDefinition(
                name="list_repositories",
                description="List repositories for the authenticated user",
                parameters={
                    "type": "object",
                    "properties": {
                        "visibility": {"type": "string", "enum": ["all", "public", "private"], "default": "all"},
                        "sort": {"type": "string", "enum": ["created", "updated", "pushed", "full_name"], "default": "updated"},
                        "direction": {"type": "string", "enum": ["asc", "desc"], "default": "desc"},
                        "per_page": {"type": "integer", "minimum": 1, "maximum": 100, "default": 30}
                    }
                },
                required_parameters=[]
            ),
            MCPToolDefinition(
                name="get_repository_content",
                description="Get content from a repository",
                parameters={
                    "type": "object",
                    "properties": {
                        "owner": {"type": "string", "description": "Repository owner"},
                        "repo": {"type": "string", "description": "Repository name"},
                        "path": {"type": "string", "description": "Path within repository", "default": ""},
                        "ref": {"type": "string", "description": "Git reference (branch, tag, commit)"}
                    }
                },
                required_parameters=["owner", "repo"]
            ),
            MCPToolDefinition(
                name="create_branch",
                description="Create a new branch in a repository",
                parameters={
                    "type": "object",
                    "properties": {
                        "owner": {"type": "string", "description": "Repository owner"},
                        "repo": {"type": "string", "description": "Repository name"},
                        "branch_name": {"type": "string", "description": "New branch name"},
                        "base_branch": {"type": "string", "description": "Base branch name"}
                    }
                },
                required_parameters=["owner", "repo", "branch_name"]
            ),
            MCPToolDefinition(
                name="commit_changes",
                description="Commit changes to a repository",
                parameters={
                    "type": "object",
                    "properties": {
                        "owner": {"type": "string", "description": "Repository owner"},
                        "repo": {"type": "string", "description": "Repository name"},
                        "branch": {"type": "string", "description": "Branch name"},
                        "message": {"type": "string", "description": "Commit message"},
                        "files": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "path": {"type": "string"},
                                    "content": {"type": "string"},
                                    "encoding": {"type": "string", "default": "utf-8"}
                                },
                                "required": ["path", "content"]
                            }
                        }
                    }
                },
                required_parameters=["owner", "repo", "branch", "message", "files"]
            ),
            MCPToolDefinition(
                name="create_pull_request",
                description="Create a pull request",
                parameters={
                    "type": "object",
                    "properties": {
                        "owner": {"type": "string", "description": "Repository owner"},
                        "repo": {"type": "string", "description": "Repository name"},
                        "title": {"type": "string", "description": "PR title"},
                        "body": {"type": "string", "description": "PR description"},
                        "head": {"type": "string", "description": "Head branch"},
                        "base": {"type": "string", "description": "Base branch"},
                        "draft": {"type": "boolean", "default": False}
                    }
                },
                required_parameters=["owner", "repo", "title", "body", "head", "base"]
            ),
            MCPToolDefinition(
                name="review_pull_request",
                description="Review a pull request",
                parameters={
                    "type": "object",
                    "properties": {
                        "owner": {"type": "string", "description": "Repository owner"},
                        "repo": {"type": "string", "description": "Repository name"},
                        "pull_number": {"type": "integer", "description": "PR number"},
                        "event": {"type": "string", "enum": ["APPROVE", "REQUEST_CHANGES", "COMMENT"]},
                        "body": {"type": "string", "description": "Review comment"}
                    }
                },
                required_parameters=["owner", "repo", "pull_number", "event", "body"]
            ),
            MCPToolDefinition(
                name="get_issues",
                description="Get repository issues",
                parameters={
                    "type": "object",
                    "properties": {
                        "owner": {"type": "string", "description": "Repository owner"},
                        "repo": {"type": "string", "description": "Repository name"},
                        "state": {"type": "string", "enum": ["open", "closed", "all"], "default": "open"},
                        "labels": {"type": "array", "items": {"type": "string"}},
                        "sort": {"type": "string", "enum": ["created", "updated", "comments"], "default": "created"}
                    }
                },
                required_parameters=["owner", "repo"]
            ),
            MCPToolDefinition(
                name="create_issue",
                description="Create a new issue",
                parameters={
                    "type": "object",
                    "properties": {
                        "owner": {"type": "string", "description": "Repository owner"},
                        "repo": {"type": "string", "description": "Repository name"},
                        "title": {"type": "string", "description": "Issue title"},
                        "body": {"type": "string", "description": "Issue description"},
                        "labels": {"type": "array", "items": {"type": "string"}},
                        "assignees": {"type": "array", "items": {"type": "string"}}
                    }
                },
                required_parameters=["owner", "repo", "title", "body"]
            )
        ]
    
    async def get_session(self, auth_config: GitHubAuthConfig) -> aiohttp.ClientSession:
        """Get or create authenticated session"""
        session_key = f"{auth_config.auth_type}_{auth_config.token[:10]}"
        
        if session_key not in self.sessions:
            headers = {
                "Accept": "application/vnd.github.v3+json",
                "User-Agent": f"{settings.app_name}/1.0"
            }
            
            if auth_config.auth_type == "personal_access_token":
                headers["Authorization"] = f"token {auth_config.token}"
            elif auth_config.auth_type == "oauth2":
                headers["Authorization"] = f"Bearer {auth_config.token}"
            
            self.sessions[session_key] = aiohttp.ClientSession(
                headers=headers,
                timeout=aiohttp.ClientTimeout(total=30)
            )
        
        return self.sessions[session_key]
    
    async def list_repositories(self, request: ListRepositoriesRequest) -> ListRepositoriesResponse:
        """List repositories for the authenticated user"""
        try:
            session = await self.get_session(request.auth_config)
            
            params = {
                "visibility": request.visibility,
                "sort": request.sort,
                "direction": request.direction,
                "per_page": request.per_page,
                "page": request.page
            }
            
            async with session.get(f"{self.base_url}/user/repos", params=params) as response:
                if response.status == 200:
                    data = await response.json()
                    repositories = [
                        GitHubRepository(
                            id=repo["id"],
                            name=repo["name"],
                            full_name=repo["full_name"],
                            private=repo["private"],
                            html_url=repo["html_url"],
                            description=repo.get("description"),
                            language=repo.get("language"),
                            default_branch=repo["default_branch"],
                            updated_at=repo["updated_at"],
                            open_issues_count=repo["open_issues_count"],
                            permissions=repo.get("permissions")
                        )
                        for repo in data
                    ]
                    
                    return ListRepositoriesResponse(
                        success=True,
                        message=f"Successfully retrieved {len(repositories)} repositories",
                        repositories=repositories
                    )
                else:
                    error_data = await response.json()
                    return ListRepositoriesResponse(
                        success=False,
                        message=f"GitHub API error: {error_data.get('message', 'Unknown error')}",
                        repositories=[]
                    )
                    
        except Exception as e:
            logger.error(f"Error listing repositories: {str(e)}")
            return ListRepositoriesResponse(
                success=False,
                message=f"Error listing repositories: {str(e)}",
                repositories=[]
            )
    
    async def get_repository_content(self, request: GetRepositoryContentRequest) -> GetRepositoryContentResponse:
        """Get repository content"""
        try:
            session = await self.get_session(request.auth_config)
            
            url = f"{self.base_url}/repos/{request.owner}/{request.repo}/contents/{request.path}"
            params = {}
            if request.ref:
                params["ref"] = request.ref
            
            async with session.get(url, params=params) as response:
                if response.status == 200:
                    data = await response.json()
                    
                    if isinstance(data, list):
                        content = [
                            GitHubContent(
                                name=item["name"],
                                path=item["path"],
                                sha=item["sha"],
                                size=item.get("size"),
                                type=item["type"],
                                download_url=item.get("download_url")
                            )
                            for item in data
                        ]
                    else:
                        content = GitHubContent(
                            name=data["name"],
                            path=data["path"],
                            sha=data["sha"],
                            size=data.get("size"),
                            type=data["type"],
                            content=data.get("content"),
                            download_url=data.get("download_url")
                        )
                    
                    return GetRepositoryContentResponse(
                        success=True,
                        message="Successfully retrieved repository content",
                        content=content
                    )
                else:
                    error_data = await response.json()
                    return GetRepositoryContentResponse(
                        success=False,
                        message=f"GitHub API error: {error_data.get('message', 'Unknown error')}",
                        content=[]
                    )
                    
        except Exception as e:
            logger.error(f"Error getting repository content: {str(e)}")
            return GetRepositoryContentResponse(
                success=False,
                message=f"Error getting repository content: {str(e)}",
                content=[]
            )
    
    async def create_branch(self, request: CreateBranchRequest) -> CreateBranchResponse:
        """Create a new branch"""
        try:
            session = await self.get_session(request.auth_config)
            
            # Get the base branch SHA
            base_branch = request.base_branch or "main"  # Default to main if not specified
            ref_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/ref/heads/{base_branch}"
            
            async with session.get(ref_url) as response:
                if response.status != 200:
                    return CreateBranchResponse(
                        success=False,
                        message=f"Could not find base branch '{base_branch}'",
                        branch_name="",
                        sha="",
                        url=""
                    )
                
                ref_data = await response.json()
                base_sha = ref_data["object"]["sha"]
            
            # Create the new branch
            create_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/refs"
            create_data = {
                "ref": f"refs/heads/{request.branch_name}",
                "sha": base_sha
            }
            
            async with session.post(create_url, json=create_data) as response:
                if response.status == 201:
                    data = await response.json()
                    return CreateBranchResponse(
                        success=True,
                        message=f"Successfully created branch '{request.branch_name}'",
                        branch_name=request.branch_name,
                        sha=data["object"]["sha"],
                        url=data["url"]
                    )
                else:
                    error_data = await response.json()
                    return CreateBranchResponse(
                        success=False,
                        message=f"GitHub API error: {error_data.get('message', 'Unknown error')}",
                        branch_name="",
                        sha="",
                        url=""
                    )
                    
        except Exception as e:
            logger.error(f"Error creating branch: {str(e)}")
            return CreateBranchResponse(
                success=False,
                message=f"Error creating branch: {str(e)}",
                branch_name="",
                sha="",
                url=""
            )
    
    async def commit_changes(self, request: CommitChangesRequest) -> CommitChangesResponse:
        """Commit changes to a repository"""
        try:
            session = await self.get_session(request.auth_config)
            
            # Get current branch SHA
            ref_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/ref/heads/{request.branch}"
            
            async with session.get(ref_url) as response:
                if response.status != 200:
                    return CommitChangesResponse(
                        success=False,
                        message=f"Could not find branch '{request.branch}'",
                        commit_sha="",
                        commit_url=""
                    )
                
                ref_data = await response.json()
                parent_sha = ref_data["object"]["sha"]
            
            # Create blobs for each file
            file_shas = []
            for file_change in request.files:
                blob_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/blobs"
                blob_data = {
                    "content": file_change.content,
                    "encoding": file_change.encoding
                }
                
                async with session.post(blob_url, json=blob_data) as blob_response:
                    if blob_response.status == 201:
                        blob_info = await blob_response.json()
                        file_shas.append({
                            "path": file_change.path,
                            "sha": blob_info["sha"]
                        })
                    else:
                        return CommitChangesResponse(
                            success=False,
                            message=f"Failed to create blob for {file_change.path}",
                            commit_sha="",
                            commit_url=""
                        )
            
            # Get base tree SHA
            commit_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/commits/{parent_sha}"
            async with session.get(commit_url) as response:
                commit_data = await response.json()
                base_tree_sha = commit_data["tree"]["sha"]
            
            # Create tree
            tree_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/trees"
            tree_data = {
                "base_tree": base_tree_sha,
                "tree": [
                    {
                        "path": file_info["path"],
                        "mode": "100644",
                        "type": "blob",
                        "sha": file_info["sha"]
                    }
                    for file_info in file_shas
                ]
            }
            
            async with session.post(tree_url, json=tree_data) as response:
                if response.status != 201:
                    return CommitChangesResponse(
                        success=False,
                        message="Failed to create tree",
                        commit_sha="",
                        commit_url=""
                    )
                
                tree_info = await response.json()
                tree_sha = tree_info["sha"]
            
            # Create commit
            commit_create_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/commits"
            commit_create_data = {
                "message": request.message,
                "tree": tree_sha,
                "parents": [parent_sha]
            }
            
            if request.author_name and request.author_email:
                commit_create_data["author"] = {
                    "name": request.author_name,
                    "email": request.author_email
                }
            
            async with session.post(commit_create_url, json=commit_create_data) as response:
                if response.status != 201:
                    error_data = await response.json()
                    return CommitChangesResponse(
                        success=False,
                        message=f"Failed to create commit: {error_data.get('message', 'Unknown error')}",
                        commit_sha="",
                        commit_url=""
                    )
                
                commit_info = await response.json()
                commit_sha = commit_info["sha"]
            
            # Update branch reference
            update_ref_url = f"{self.base_url}/repos/{request.owner}/{request.repo}/git/refs/heads/{request.branch}"
            update_ref_data = {"sha": commit_sha}
            
            async with session.patch(update_ref_url, json=update_ref_data) as response:
                if response.status == 200:
                    return CommitChangesResponse(
                        success=True,
                        message=f"Successfully committed {len(request.files)} files",
                        commit_sha=commit_sha,
                        commit_url=commit_info["html_url"]
                    )
                else:
                    return CommitChangesResponse(
                        success=False,
                        message="Failed to update branch reference",
                        commit_sha="",
                        commit_url=""
                    )
                    
        except Exception as e:
            logger.error(f"Error committing changes: {str(e)}")
            return CommitChangesResponse(
                success=False,
                message=f"Error committing changes: {str(e)}",
                commit_sha="",
                commit_url=""
            )
    
    async def create_pull_request(self, request: CreatePullRequestRequest) -> CreatePullRequestResponse:
        """Create a pull request"""
        try:
            session = await self.get_session(request.auth_config)
            
            url = f"{self.base_url}/repos/{request.owner}/{request.repo}/pulls"
            data = {
                "title": request.title,
                "body": request.body,
                "head": request.head,
                "base": request.base,
                "draft": request.draft,
                "maintainer_can_modify": request.maintainer_can_modify
            }
            
            async with session.post(url, json=data) as response:
                if response.status == 201:
                    pr_data = await response.json()
                    pull_request = PullRequestModel(
                        id=pr_data["id"],
                        number=pr_data["number"],
                        title=pr_data["title"],
                        body=pr_data.get("body"),
                        state=pr_data["state"],
                        html_url=pr_data["html_url"],
                        created_at=pr_data["created_at"],
                        updated_at=pr_data["updated_at"],
                        head=pr_data["head"],
                        base=pr_data["base"],
                        mergeable=pr_data.get("mergeable"),
                        merged=pr_data["merged"],
                        draft=pr_data["draft"]
                    )
                    
                    return CreatePullRequestResponse(
                        success=True,
                        message=f"Successfully created PR #{pr_data['number']}",
                        pull_request=pull_request
                    )
                else:
                    error_data = await response.json()
                    return CreatePullRequestResponse(
                        success=False,
                        message=f"GitHub API error: {error_data.get('message', 'Unknown error')}",
                        pull_request=PullRequestModel(
                            id=0, number=0, title="", state="", html_url="",
                            created_at="", updated_at="", head={}, base={},
                            merged=False, draft=False
                        )
                    )
                    
        except Exception as e:
            logger.error(f"Error creating pull request: {str(e)}")
            return CreatePullRequestResponse(
                success=False,
                message=f"Error creating pull request: {str(e)}",
                pull_request=PullRequestModel(
                    id=0, number=0, title="", state="", html_url="",
                    created_at="", updated_at="", head={}, base={},
                    merged=False, draft=False
                )
            )
    
    async def get_mcp_tools(self) -> List[MCPToolDefinition]:
        """Get MCP tool definitions for Claude integration"""
        return self.mcp_tools
    
    async def execute_mcp_operation(self, operation: GitHubMCPOperationType, params: Dict[str, Any], auth_config: GitHubAuthConfig) -> Dict[str, Any]:
        """Execute MCP operation and return structured result"""
        try:
            if operation == GitHubMCPOperationType.LIST_REPOSITORIES:
                request = ListRepositoriesRequest(auth_config=auth_config, **params)
                response = await self.list_repositories(request)
                return response.dict()
            
            elif operation == GitHubMCPOperationType.GET_REPOSITORY_CONTENT:
                request = GetRepositoryContentRequest(auth_config=auth_config, **params)
                response = await self.get_repository_content(request)
                return response.dict()
            
            elif operation == GitHubMCPOperationType.CREATE_BRANCH:
                request = CreateBranchRequest(auth_config=auth_config, **params)
                response = await self.create_branch(request)
                return response.dict()
            
            elif operation == GitHubMCPOperationType.COMMIT_CHANGES:
                request = CommitChangesRequest(auth_config=auth_config, **params)
                response = await self.commit_changes(request)
                return response.dict()
            
            elif operation == GitHubMCPOperationType.CREATE_PULL_REQUEST:
                request = CreatePullRequestRequest(auth_config=auth_config, **params)
                response = await self.create_pull_request(request)
                return response.dict()
            
            else:
                return {
                    "success": False,
                    "message": f"Operation {operation} not implemented yet"
                }
                
        except Exception as e:
            logger.error(f"Error executing MCP operation {operation}: {str(e)}")
            return {
                "success": False,
                "message": f"Error executing operation: {str(e)}"
            }
    
    async def cleanup(self):
        """Cleanup all sessions"""
        for session in self.sessions.values():
            await session.close()
        self.sessions.clear()


# Global service instance
github_mcp_service = GitHubMCPService() 